File: src/test/baekjoon/python/binarysearch/CutTree.py
N, M = map(int, input().split())
trees = list(map(int, input().split()))
start = 1
end = max(trees)
def binary_search(arr, start, end):
    while start <= end:
        m_tree = 0
        mid = (start + end) // 2
        for x in arr:
            if x >= mid:
                m_tree += x - mid
        if m_tree >= M:
            start = mid + 1
        else:
            end = mid - 1
    return end
print(binary_search(trees, start, end))

File: src/test/baekjoon/java/binarysearch/CutTree.java
package test.baekjoon.java.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class CutTree {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] values = br.readLine().split(" ");
        int N = Integer.parseInt(values[0]);
        int M = Integer.parseInt(values[1]);
        int[] trees = Arrays.stream(br.readLine().split(" ")).mapToInt(i -> Integer.parseInt(i)).toArray();
        int start = 1;
        int end = 0;
        for(int tree : trees) {
            if(end < tree) {
                end = tree;
            }
        }

        while(start <= end) {
            long m_tree = 0;
            int mid = (start + end) / 2;
            for(int tree : trees) {
                if(tree >= mid) {
                    m_tree += tree - mid;
                }
            }
            if(m_tree >= M) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }

        System.out.println(end);
    }
}
